### DISCUSSION TITLE
AWS CloudHSM JCE provider & JJWT

### DISCUSSION THREAD

**Discussion by mnylen** [jwtk/jjwt#934] [Q&A] [answered]
Hello!

I've been exploring the use of AWS CloudHSM for signing and verifying JWTs. This works pretty easily with RS256 and ES256 algorithms, but HS256 is proving to be a bit difficult, because the `SecretKey` instance returned from HSM just has `GenericSecret` as algorithm, which JJWT rejects:

```
io.jsonwebtoken.security.SignatureException: Unable to compute HS256 signature. Cause: The signing key's algorithm 'GenericSecret' does not equal a valid HmacSHA* algorithm name or PKCS12 OID and cannot be used with HS256.
```

The same key however works well when used directly with `javax.crypto.Mac` manually.

I'm wondering if there's a way to tell JJWT that the key has `HmacSHA256` algorithm, or somehow force it to accept the key? As far as I know, as long as the key is >= 32 bytes, it should be suitable, as long as the provider can calculate the mac using the key?

```
var provider = new CloudHsmProvider();
Security.addProvider(provider);

var keystore = KeyStore.getInstance(CloudHsmProvider.PROVIDER_NAME);
keystore.load(null, null);

var key = (SecretKey)keystore.getKey("hmac-key", null);

// this works
var mac = Mac.getInstance("HmacSHA256", CloudHsmProvider.PROVIDER_NAME);
mac.init(key);
mac.update("hello".getBytes());
System.out.println("digest: " + Base64.getEncoder().encodeToString(mac.doFinal()));

// this throws io.jsonwebtoken.security.SignatureException: Unable to compute HS256 signature
var token = Jwts.builder().subject("hello").signWith(key, Jwts.SIG.HS256).compact();
```

Edit: My own replies to this thread deviated slightly to also include problems encountered with AES-GCM encryption, so renamed the discussion from being specific to the HMAC keys to just be generally AWS CloudHSM & JJWT related. :)

**Comment by mnylen** [comment]
It seems this error is coming from https://github.com/jwtk/jjwt/blob/c673b76ef5b74ddaf51b4e80fc632ddb62b7e29a/impl/src/main/java/io/jsonwebtoken/impl/security/DefaultMacAlgorithm.java#L143 where there seems to be some support for generic secrets, but I guess the CloudHSM `GenericSecretKey` variant doesn't get matched.

The actual key class is `com.amazonaws.cloudhsm.jce.provider.GenericSecretKey` and the `getAlgorithm()` returns `GenericSecret`

Would JJWT implementation need to be extended to support this, or is there a way to convert so that the algorithm would match the expected one?

**Reply by mnylen** [reply]
The only workaround I was able to find was to "monkey-patch" the whole `DefaultMacAlgorithm` to not validate the algorithm name. But after that, the HS256 sign/verify works as expected with CloudHSM keys.

First, I tried to create a custom `ProviderKey` implementation that overrides the `getAlgorithm()` method, but unfortunately it seems the `ProviderKey` is unwrapped by the time `DefaultMacAlgorithm`is used. Although still hacky, it would've isolated addition instead of replacing an existing class.

**Reply by mnylen** [reply]
I created a PR to allow bypassing the algorithm name validation for any `SecretKey` where `getAlgorithm()` returns `"GenericSecret"`. https://github.com/jwtk/jjwt/pull/935

**Comment by mnylen** [comment]
Not exactly related to HMAC keys, but JWEs with A256GCM encryption can't use AES keys retrieved via CloudHSM JCE Provider, as they report key length to be 0.

Trying to supply `com.amazonaws.cloudhsm.jce.provider.AesKey` to `encryptWith` gives

```
io.jsonwebtoken.security.WeakKeyException: The 'A256GCM' algorithm requires keys with a length of 256 bits (32 bytes).  The provided key has a length of 0 bits (0 bytes).
```

which seems to be thrown from here https://github.com/jwtk/jjwt/blob/master/impl/src/main/java/io/jsonwebtoken/impl/security/AesAlgorithm.java#L98

The problem is that `getEncoded()` for `com.amazonaws.cloudhsm.jce.provider.AesKey` returns null (as the key is not extractable), which `Bytes.bitLength` assumes to mean 0 bits. 🤔  Is this how `null` values are intended to be interpreted?

**Reply by mnylen** [reply]
According to https://docs.oracle.com/javase/8/docs/api/java/security/Key.html#getEncoded-- the `getEncoded()` can return `null` if the key does not support encoding, so it seems that jjwt should maybe treat this as "length unknown" case, and skip the key length validation?

And FWIW, the `DefaultMacAlgorithm` (that the original question was related to) uses `KeysBridge.findBitLength` to determine the key length (https://github.com/jwtk/jjwt/blob/c673b76ef5b74ddaf51b4e80fc632ddb62b7e29a/impl/src/main/java/io/jsonwebtoken/impl/security/KeysBridge.java#L110). This method seems to handle the case where `getFormat()` returns `null`. Is there something specific about `AesAlgorithm` that it can't use this method that would already correctly handle this case, and instead has it's own length validation?

**Reply by mnylen** [reply]
Continuing with the monkey-patch theme, having dealt with the key length issue in a crude way, the next problem is that CloudHSM JCE provider expects the IV to be all zeros, as CloudHSM generates it's own IV. 🤔 

From https://docs.aws.amazon.com/cloudhsm/latest/userguide/pkcs11-mechanisms.html

> [1] When performing AES-GCM encryption, the HSM does not accept initialization vector (IV) data from the application. You must use an IV that it generates. The 12-byte IV provided by the HSM is written into the memory reference pointed to by the pIV element of the CK_GCM_PARAMS parameters structure that you supply. To prevent user confusion, PKCS #11 SDK in version 1.1.1 and later ensures that pIV points to a zeroized buffer when AES-GCM encryption is initialized.

The exception thrown is:

```
io.jsonwebtoken.security.SecurityException: Cipher callback execution failed: Supplied IV is invalid. Reason: Iv invalid. Reason: Expect IV to be all 0
	at io.jsonwebtoken.impl.security.JcaTemplate.execute(JcaTemplate.java:130)
	at io.jsonwebtoken.impl.security.JcaTemplate.execute(JcaTemplate.java:135)
	at io.jsonwebtoken.impl.security.JcaTemplate.withCipher(JcaTemplate.java:164)
	at io.jsonwebtoken.impl.security.GcmAesAeadAlgorithm.encrypt(GcmAesAeadAlgorithm.java:60)
	at io.jsonwebtoken.impl.DefaultJwtBuilder$3.apply(DefaultJwtBuilder.java:680)
	at io.jsonwebtoken.impl.lang.DelegatingCheckedFunction.apply(DelegatingCheckedFunction.java:28)
	at io.jsonwebtoken.impl.lang.PropagatingExceptionFunction.apply(PropagatingExceptionFunction.java:57)
	at io.jsonwebtoken.impl.DefaultJwtBuilder.encrypt(DefaultJwtBuilder.java:684)
	at io.jsonwebtoken.impl.DefaultJwtBuilder.encrypt(DefaultJwtBuilder.java:739)
	at io.jsonwebtoken.impl.DefaultJwtBuilder.compact(DefaultJwtBuilder.java:531)
	at Main.aesGcmEncryptDecrypt(Main.java:49)
	at Main.main(Main.java:25)
Caused by: com.amazonaws.cloudhsm.jce.jni.exception.InvalidIvException: Supplied IV is invalid. Reason: Iv invalid. Reason: Expect IV to be all 0
	at com.amazonaws.cloudhsm.jce.jni.Session.do_validateEncryptAesGcm(Native Method)
	at com.amazonaws.cloudhsm.jce.jni.Session.validateEncryptAesGcm(Session.java:190)
	at com.amazonaws.cloudhsm.jce.provider.AesGcmCipher.validateKey(AesGcmCipher.java:426)
	at com.amazonaws.cloudhsm.jce.provider.CloudHsmCipherBase.setKey(CloudHsmCipherBase.java:573)
	at com.amazonaws.cloudhsm.jce.provider.CloudHsmCipherBase.engineInit(CloudHsmCipherBase.java:188)
	at com.amazonaws.cloudhsm.jce.provider.AesGcmCipher.engineInit(AesGcmCipher.java:204)
	at java.base/javax.crypto.Cipher.implInit(Cipher.java:843)
	at java.base/javax.crypto.Cipher.chooseProvider(Cipher.java:901)
	at java.base/javax.crypto.Cipher.init(Cipher.java:1433)
	at java.base/javax.crypto.Cipher.init(Cipher.java:1364)
	at io.jsonwebtoken.impl.security.GcmAesAeadAlgorithm$1.apply(GcmAesAeadAlgorithm.java:63)
	at io.jsonwebtoken.impl.security.GcmAesAeadAlgorithm$1.apply(GcmAesAeadAlgorithm.java:60)
	at io.jsonwebtoken.impl.security.JcaTemplate.execute(JcaTemplate.java:120)
	at io.jsonwebtoken.impl.security.JcaTemplate.access$100(JcaTemplate.java:65)
	at io.jsonwebtoken.impl.security.JcaTemplate$2.get(JcaTemplate.java:138)
	at io.jsonwebtoken.impl.security.JcaTemplate.execute(JcaTemplate.java:125)
```

**Reply by mnylen** [reply]
So the following modifications are needed to GcmAesAeadAlgorithm to support CloudHSM:

* the `iv` must be initialized as byte array with length 12 and with only zero values
* after `cipher.doFinal()`, need to retrieve the IV CloudHSM generated with `cipher.getIV()`

**Reply by mnylen** [reply]
Okay, so the key length issue aside, I think using provider-generated IV will require the use of custom encryption algorithm. It would be nice if the custom alg could still wrap the `GcmAesAlgorithm` and just pass it `AeadRequest` that implements `IvProvider` that returns the zero-IV.

That would be enough for the encryption, but for decryption, the `GcmAesAlgorithm` would need to be modified so that it could read the actual IV from `Cipher`. Wonder if this would be safe to do always, or if there should be a special `ProviderGeneratedIv` interface that the request could implement. 🤔

**Reply by mnylen** [reply]
Would something like this be acceptable for `GcmAesAeadAlgorithm#encrypt()`:

```java
    @Override
    public void encrypt(final AeadRequest req, AeadResult res) throws SecurityException {

        Assert.notNull(req, "Request cannot be null.");
        Assert.notNull(res, "Result cannot be null.");

        final SecretKey key = assertKey(req.getKey());
        final InputStream plaintext = Assert.notNull(req.getPayload(),
                "Request content (plaintext) InputStream cannot be null.");
        final OutputStream out = Assert.notNull(res.getOutputStream(),
                "Result ciphertext OutputStream cannot be null.");
        final InputStream aad = req.getAssociatedData();
        final byte[] ivSpecIv = ensureInitializationVector(req);
        final AlgorithmParameterSpec ivSpec = getIvSpec(ivSpecIv);

        final TagAndIv result = jca(req).withCipher(new CheckedFunction<Cipher, TagAndIv>() {
            @Override
            public TagAndIv apply(Cipher cipher) throws Exception {
                cipher.init(Cipher.ENCRYPT_MODE, key, ivSpec);
                byte[] taggedCiphertext = withCipher(cipher, plaintext, aad, out);
                // When using GCM mode, the JDK appends the authentication tag to the ciphertext, so let's extract it:
                // (tag has a length of BLOCK_BYTE_SIZE):
                int ciphertextLength = Bytes.length(taggedCiphertext) - BLOCK_BYTE_SIZE;
                Streams.write(out, taggedCiphertext, 0, ciphertextLength, "Ciphertext write failure.");
                byte[] tag = new byte[BLOCK_BYTE_SIZE];
                System.arraycopy(taggedCiphertext, ciphertextLength, tag, 0, BLOCK_BYTE_SIZE);

                return new TagAndIv(tag, cipher.getIV());
            }
        });
        Streams.flush(out);
        Streams.reset(plaintext);

        res.setTag(result.getTag()).setIv(result.getIv());
    }
```

The modifications to the original would be:

1. Return a new value class TagAndIv from `withCipher(...)` block
2. The actual IV is read from `Cipher`

There probably needs to be assertion that `cipher.getIV()` doesn't return null.

And the idea here would be that a custom enc algorithm could pass `IvSupplier` with the zero-IV, so the cipher would be initialized with that, but the actual IV will be read from the cipher.

Similar changes should be done to other algorithms as well, but before I start the work, would like some direction. :)

**Reply by lhazlewood** [reply]
> And FWIW, the `DefaultMacAlgorithm` (that the original question was related to) uses `KeysBridge.findBitLength` to determine the key length (https://github.com/jwtk/jjwt/blob/c673b76ef5b74ddaf51b4e80fc632ddb62b7e29a/impl/src/main/java/io/jsonwebtoken/impl/security/KeysBridge.java#L110). This method seems to handle the case where `getFormat()` returns `null`. Is there something specific about `AesAlgorithm` that it can't use this method that would already correctly handle this case, and instead has it's own length validation?

Nope, nothing specific about `AesAlgorithm` other than it was implemented before `KeysBridge#findBitLength` was written 😄.  It definitely makes sense to refactor `AesAlgorithm` to use the same logic for the same reasons.

**Reply by lhazlewood** [reply]
A few things here about the `GcmAesAeadAlgorithm` implementation and initialization vectors:

The `GcmAesAeadAlgorithm` is in the `impl` package which is the only JJWT `.jar` that does not adhere to semantic versioning per https://github.com/jwtk/jjwt?tab=readme-ov-file#understanding-jjwt-dependencies.  When JDK 9 modules are supported via `module-info.java` files, internal implementations won't be accessible at all, so it's best not to depend on anything in that .jar. (this https://github.com/jwtk/jjwt/discussions/929#discussioncomment-8975616 might be helpful as well).

So the 'safer' immediate fix here is implement the `AeadAlgorithm` interface directly to do whatever you need and override the default implementation by using `JwtParserBuilder#enc()` (see that method's JavaDoc).


Regarding a zeroed-out IV, I have to admit that feels _really_ wrong and a very poor design decision on the part of the AWS team.  Proper IVs are _never_ deterministic - forcing an all-zero IV could have catastrophic affects on an application where the wrong IV is used (e.g. IV is zeroed out but CloudHSM is not used in that context, and now there's no random initialization destroying the application's security).  That AWS would require this is irresponsible IMO, it'd be better to NOT provide an IV at all when initializing the Cipher to clearly indicate that the Cipher is expected to generate one automatically.

So it seems that some thought needs to be put into how IVs are generated within the `AesAlgorithm` implementation and subclasses.  The current implementation works well but it's not extensible for things like AWS's oddity.  If we can guarantee that an IV will be securely generated at all times by the Cipher, there's no need for JJWT to generate one itself, and we can remove that code.  Testing/verification and non-null/non-empty assertions would need to be put in the code to guarantee this of course.

> There probably needs to be assertion that cipher.getIV() doesn't return null.

Yep, non-null but also has the exact required AES IV length (96 bits for GCM mode, 128 bits otherwise).

In the meantime of course, a custom `AeadAlgorithm` implementation avoids any delay waiting for a JJWT release.

**Reply by mnylen** [reply]
Yeah, I agree that it'd be quite bad if the IV then isn't generated. So perhaps the right way for all of these issues would be to just "implement" (copy-paste with some surgery) our own algorithms and assert that CloudHSM provider is used in them.

It's probably hard to guarantee that all `Cipher` implementations would generate the IV securely. It would need to be some special interface that the request implements, like `ProviderGeneratedIv`. But then this would also need to provide the IV that the `Cipher` gets initialized with, as it's unlikely that other HSMs would similarly except zero-IV. But with that approach, one could implement their own algorithm that just delegates a modified request to `Jwts.ENC.A256GCM`.

**Reply by mnylen** [reply]
Having done some further testing, it seems with CloudHSM, if you call `Cipher#init()` without the third `params` object, it also generates the IV for you. So no need for the zero-IV.

`null` or empty byte array won't be accepted by `GCMParametersSpec`

**Comment by lhazlewood** [comment, marked as answer]
Hi @mnylen!

I was out over the weekend and yesterday was as super busy day, so we haven't been able to review these comments in detail until now.  Please see below:

> I've been exploring the use of AWS CloudHSM for signing and verifying JWTs. This works pretty easily with RS256 and ES256 algorithms, but HS256 is proving to be a bit difficult, because the `SecretKey` instance returned from HSM just has `GenericSecret` as algorithm, which JJWT rejects:
> 
> ```
> io.jsonwebtoken.security.SignatureException: Unable to compute HS256 signature. Cause: The signing key's algorithm 'GenericSecret' does not equal a valid HmacSHA* algorithm name or PKCS12 OID and cannot be used with HS256.
> ```
> 
> The same key however works well when used directly with `javax.crypto.Mac` manually.

The `HmacSHA*` name and OID checks are in place to prevent people from accidentally using keys that they shouldn't be using for `Mac` calculations.  This happens _all the time_ in applications: where one key is used for one cryptographic operation and then used again for an entirely different operation (e.g. Mac and AES encryption, or using the same RSA key pair for both encryption and signatures), and this is a well-documented poor practice that compromises key security and most application developers are never aware of the problem.  So the name check helps ensure that someone doesn't accidentally use an AES key (for example) as a Mac key.

The JCA doesn't do any key algorithm name checks (as you've shown in your code example), but the JCA has many problems like this that are poor practices that I won't get into now 😓 .  So JJWT tries to 'do better' and avoid these scenarios entirely.

It's unfortunate that Sun's own PKCS11 keys, and now evidently AWS's, don't make this distinction.  But I think we can make an exception for Cloud HSM as well.  For example, we have this check:

https://github.com/jwtk/jjwt/blob/c673b76ef5b74ddaf51b4e80fc632ddb62b7e29a/impl/src/main/java/io/jsonwebtoken/impl/security/KeysBridge.java#L98-L102

We could probably add something similar and then wrap both checks in a utility method that covers both cases for use in various algorithms.

> ```java
> var provider = new CloudHsmProvider();
> Security.addProvider(provider);
> ```

JJWT has been tested extensively when using Sun's PKCS11 provider, which acts as a bridge to the native HSM implementation.  Have you tried using Cloud HSM with the [SunPKCS11](https://docs.oracle.com/en/java/javase/20/security/pkcs11-reference-guide1.html#GUID-6DA72F34-6C6A-4F7D-ADBA-5811576A9331) provider instead of directly as shown in the code snippet above?  If this works, it may likely solve the issues you're seeing.

**Reply by mnylen** [reply]
I could maybe try that. Just wonder if the SUNPKCS11 implements all the loadbalancing bits over multiple HSMs that the CloudHSM one does.

~But AES-GCM will require the IV to be generated on the HSM, see the other thread in this discussion.~ Ah, nevermind - saw that you answered that already.
